### PR TITLE
Curie refactor

### PR DESCRIPTION
**Pull request by markquintontulloch**
As discussed at the DQM/Modelling meeting on 21st Nov

### CODE REVIEW

**Review by oblodgett — COMMENTED** [review]
I would have guessed this refactor would have been more than this?

**Comment by markquintontulloch** [comment]
> I would have guessed this refactor would have been more than this?

I don't think so, although I may well have missed something.  Most things with a curie, mod_entity_id, or mod_internal_id inherit from something else, so it's only a few parent classes that are changing.  The DTO classes are largely unaffected.

The Java side on the other hand..... well, the migration will be fun!

File: model/schema/core.yaml [hunk 1]
```diff
@@ -82,28 +82,65 @@ classes:
...
+  SubmittedCurieObject:
+    is_a: CurieAuditedObject
+    description: >-
+      Parent class for all LinkML classes of submitted entities or annotations
+      that have (or will have) a curie generated at the Alliance
+    slots:
+      - mod_entity_id
+      - mod_internal_id
+    rules:
+      - postconditions:
```
Comment (chris-grove) [inline model/schema/core.yaml:101]: @markquintontulloch 
We need to consider the fact that currently the DiseaseAnnotation class has fields for curie, mod_entity_id, and mod_internal_id but SGD, ZFIN, RGD, and OMIM (via RGD) do not provide any of those fields. We only have a generated internal UniqueID. This was a requirement of disease annotation submission discussed early on.

Comment (markquintontulloch) [reply model/schema/core.yaml:101]: Sure, will remove the rules to cover that case

**Comment by chris-grove** [comment]
@markquintontulloch @oblodgett 
I think the challenging bit here is that we still have (and can likely expect to have) additional classes that have curies but inherit from classes that do not (and arguably should not always) have a curie. For example, the following classes still have curie slots:
```
Laboratory
GeneNomenclatureSet
Identifier
GeneInteraction
GeneInteractionDTO
Annotation
```
It would take considerable reconfiguration of the model to make it so that these classes all inherit from a chain of ancestors that all have a curie. We could opt to add a curie field in some classes that do not inherit from CurieAuditedObject, but then we would modify the definition of CurieAuditedObject so that it is not declared to be the parent/root class of ALL classes with a curie. What do you think?

**Comment by markquintontulloch** [comment]
@chris-grove I may be wrong (@oblodgett can correct me), but I don't think the issue with Hibernate would prevent us from adding the curie field to a child class when the parent doesn't have a curie.  As all of the entities would have the same base class (AuditedObject), it wouldn't matter where in the hierarchy curie was implemented.  The issue now is that in Java the base class of Construct is Reagent (it inherits from AuditedObject but this isn't implemented as a table), while the base class of genomic entities is BiologicalEntity.

**Review by chris-grove — APPROVED** [review]

**Comment by markquintontulloch** [comment]
@chris-grove Also moved `data_provider` from `Reagent` and `BiologicalEntity` into `SubmittedObject` - I think we will want this for all submitted classes.

**Review by sierra-moxon — APPROVED** [review]

**Pull request merged**